Update QC report to include post processing stats and UMAP

allyhawkins opened the pull request:
Closes #121 

Here I started making changes to the QC report to incorporate the additional processed SCE object that we are now creating as output in `scpca-nf`. I added a param to the report for the `processed_sce` and set the default to `NULL`, similar to the setup for the `filtered_sce`. I also added to the `Cell Statistics` table additional rows to include the method used for filtering low quality cells, number of cells, and normalization method. I also thought we should include any parameters that might be altered such as the min genes cutoff and the probability of a cell being compromised cutoff used for miQC but we currently don't store those in the metadata of the SCE object and we probably should. I still added a line for them in the table here and if we like that change we can file an issue to add that change to the script in `scpca-nf`. 

The other thing I did was add a UMAP if the `processed_sce` param was not null. I wasn't quite sure what we wanted for color here so I tried a few things. I think if we aren't going to be coloring by clusters then we probably want to color by either total UMI count or by total genes detected which is what I did for now. I did also play around with plotting the gene expression of the most highly variable gene too and that might be promising. I included a screenshot here: 

![image](https://user-images.githubusercontent.com/54039191/194370704-d3444f3d-a65e-4d5e-bfa2-9dd959ce96fe.png)

I think if we went this route we would want to also include the gene symbol, but I didn't do much formatting since I wasn't sure if that was really a joke or not... We could also plot a few of the most HVGs and make a faceted UMAP... 

Another thought I had was that we might want to include a plot showing the filtering cutoffs of the cells that were kept vs. thrown out after using miQC + min gene threshold. I didn't add it in yet because I think that might be a separate issue, but just wanted to mention that here. 

I'm attaching a current copy of the QC report with the new additions and filing this as a draft PR just to get some initial feedback on the UMAP mostly. Once we decide the direction we want to go with the UMAP then I'll polish up the plot and report and add in a description around it like we have for the other plots. 
[test.html.zip](https://github.com/AlexsLemonade/scpcaTools/files/9726903/test.html.zip)

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -204,17 +207,39 @@ knitr::kable(processing_info, align = 'r') |>
+      "Method used to filter low quality cells" = format(processed_meta$ccdl_filter_method),
+      "Cells after filtering low quality cells" = format(dim(processed_sce)[2], big.mark = ',', scientific = FALSE),
+      "Normalization method"                    = format(processed_meta$normalization),
+      "Minimum genes per cell cutoff"           = format(processed_meta$min_gene_cutoff)
```
I noticed in the report that this was NA, but this should usually be set, correct? Or are we not using a minimum even with miQC?

allyhawkins replied:
this is usually set but we haven't added it to the metadata yet, so we need to add it to the metadata of the SCE object for it to print the value here. That needs to happen in `scpca-nf` though and not here.

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -286,7 +311,7 @@ ggplot(grouped_celldata, aes(x = med_rank, y = med_sum, color = pct_passed)) +
 
 The total UMI count of each droplet (barcode) plotted against the rank of that droplet allows visualization of the distribution of sequencing depth across droplets. 
-The droplets that are expected to contain cells were identified with [`DropletUtils::emptyDrops()`](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html), which uses both the total UMI counts and expressed gene content ([Lun  _et al._ 2019](https://doi.org/10.1186/s13059-019-1662-y)).
+The droplets that are expected to contain cells were identified with [`DropletUtils::emptyDropsCellRanger()`](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html), which uses both the total UMI counts and expressed gene content ([Lun  _et al._ 2019](https://doi.org/10.1186/s13059-019-1662-y)).
```
Do we want to say something like "unless otherwise specified"?

I think we may also have to update/add the citation to the alevin-fry paper for the details of that method? Or say something about how it derived from the Lun et al method.

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -356,6 +381,41 @@ This can be the result of a low-quality library or may occur if there is no mito
+                         point_size = 0.1, 
+                         colour_by = "detected") +
+    theme_bw() + 
+    labs(
+      color = "Number of genes detected"
+    )
```
With `plotReducedDim` this doesn't work, for some reason. We have to go back to

```suggestion
    guides(
      color = guide_colorbar(title="Number of genes detected"))
    )
```

unless you can find some way around it!

We probably want to break that legend into two lines with a careful `\n`

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -356,6 +381,41 @@ This can be the result of a low-quality library or may occur if there is no mito
+  }
+  
+  # create UMAP colored by number of genes detected 
+  scater::plotReducedDim(processed_sce, 
+                         dimred = "UMAP", 
```
Just simplifying a bit since this is a default `UMAP`
```suggestion
  scater::plotUMAP(processed_sce, 
```

jashapiro replied:
Note that this change wasn't just removing `dimred`, but also changing to `plotUMAP()`, which removes the need for that argument (as long as there is a reduced dimension table named UMAP).

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -356,6 +381,41 @@ This can be the result of a low-quality library or may occur if there is no mito
+  # create UMAP colored by number of genes detected 
+  scater::plotReducedDim(processed_sce, 
+                         dimred = "UMAP", 
+                         point_size = 0.1, 
```
Maybe a bit bigger?
```suggestion
                         point_size = 0.3, 
```

jashapiro reviewed (commented):
This looks like a good start to me! I think the plot looks pretty good as it is, but of course there are places for improvement!

> I think if we went this route we would want to also include the gene symbol, but I didn't do much formatting since I wasn't sure if that was really a joke or not... We could also plot a few of the most HVGs and make a faceted UMAP...

I like this initial plot of the gene counts; it adds a nice bit of color that does have some functional information in it. I think it is a good choice until we add clustering, or in the absence of clustering.

I do also like the idea of faceting a few different genes as a secondary view, but I would probably end up wanting to do that more manually than with the `plotUMAP` function, so I might save it for another day/PR. I would implement this as our own function, which is easier given that we are already working within a package! 

(Briefly, I would join the UMAP matrix to the expression table with the genes of interest, `pivot_longer` to put all the expression values into a single column with a separate column of gene names, then plot with `facet_wrap`)

> Another thought I had was that we might want to include a plot showing the filtering cutoffs of the cells that were kept vs. thrown out after using miQC + min gene threshold. I didn't add it in yet because I think that might be a separate issue, but just wanted to mention that here.

I kind of like this idea, but what kind of plot were you thinking of? Was this going to be some kind of distribution plot, or would we overlay it one of the other plots we already have somehow?

allyhawkins commented:
> I do also like the idea of faceting a few different genes as a secondary view, but I would probably end up wanting to do that more manually than with the plotUMAP function, so I might save it for another day/PR. I would implement this as our own function, which is easier given that we are already working within a package!

This sounds good to me! I'll leave the faceted plots showing HVGs for another PR and file an issue about adding that. 

> I kind of like this idea, but what kind of plot were you thinking of? Was this going to be some kind of distribution plot, or would we overlay it one of the other plots we already have somehow?

I was thinking something similar to this plot where we show both distributions and then we just label the `keep` cells. But here we would only have one dashed line indicating the num genes detected cutoff. Would also want to file an issue about this to do in a separate PR.

![image](https://user-images.githubusercontent.com/54039191/194624070-144a47b1-56d9-4057-a2f1-b13a5c761999.png)

allyhawkins commented:
Thanks for taking a look at this! I went ahead and added a section for the UMAP and a brief description. I was struggling with a heading title and the detail in the description, so any feedback on that would be great! 

I also added a "unless otherwise specified" to the filtering statement and a "adapted from" in the reference. These were the lines from the alevin-fry paper, about where they got `emptyDropsCellRanger` from: 

> To compare the results from different quantification tools in the snRNA-seq setting under a consistent and robust barcode filtering approach, we implemented the emptyDrops_CR functionality of STARsolo in R. The emptyDrops_CR filtering method is, itself, reverse engineered from the hybrid filtering strategy of Cell Ranger, which combines filtering based on various thresholds with the statistical testing method introduced by Lun et al. ([13](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8933848/#R40)). This functionality is now included in the DropletUtils R package as the emptyDropsCellRanger function.

I also made the small adjustments to the UMAP, but I couldn't remove the `dimred` argument because it throws an error that it's required. I also thought about using another color scale but we would ned a colorblind friendly continuous scale? Any suggestions on other ones besides viridis that you like? 

I filed the other issues that will need to be addressed too. 

Attached is the most updated copy of the report. 
[SCPCL000001_test.html.zip](https://github.com/AlexsLemonade/scpcaTools/files/9737020/SCPCL000001_test.html.zip)

jashapiro reviewed (approved):
LGTM! My only comment below is about the name of the section: For some reason "Dimensionality Reduction" seems a bit opaque to me. I don't feel very strongly about this at all.

allyhawkins commented:
> LGTM! My only comment below is about the name of the section: For some reason "Dimensionality Reduction" seems a bit opaque to me. I don't feel very strongly about this at all.

I would agree that it's vague. I wasn't really sure what to name it... But "UMAP" also seemed opaque. I also think we might update this as we add in additional UMAPs showing HVGs. If you have other ideas I am definitely open to them.
